[android] Refine tunnel mode video playback

chiyuanchan opened the pull request:
We modify some seek/preroll logic using OnFirstTunnelFrameReadyListener and BUFFER_FLAG_DECODE_ONLY.

Change-Id: I056279784b22b8174ec636573756a3bc6e258f96

b/244649793

chiyuanchan committed:
[android] Refine video preroll condition for tunnel mode

We modify some seek/preroll logic using OnFirstTunnelFrameReadyListener and BUFFER_FLAG_DECODE_ONLY.

Change-Id: Ie3b489b9627e4ca0588c14b06ff3a2b5805ecc4c

billy0059 commented:
Hi @jasonzhangxx, Per discussed, MTK has removed "variable speed playback" feature for the tunnel mode.  
Please take a look first and see if there are some problems.  

We should also start to think about what tests to conduct to make sure the change is good on Android 14 platforms.

jasonzhangxx commented on starboard/android/shared/media_decoder.h:
```diff
@@ -73,6 +74,7 @@ class MediaDecoder final
     // MediaCodecBridge so the host can have a chance to do necessary cleanups
     // before the MediaCodecBridge is flushed.
     virtual void OnFlushing() = 0;
+    virtual int64_t seek_to_time() = 0;
```
I don't think it's a good idea to add `seek_to_time()` here, but let's keep it as is, so that we can keep the experiment change to tunnel mode only.

jasonzhangxx reviewed (commented):
Please ensure that all changes only affect tunnel mode pipeline.

Jous-Chan commented on starboard/android/apk/app/src/main/java/dev/cobalt/media/MediaCodecBridge.java:
```diff
@@ -531,6 +532,24 @@ public void onFrameRendered(MediaCodec codec, long presentationTimeUs, long nano
+          new MediaCodec.OnFirstTunnelFrameReadyListener() {
+            @Override
+            public void onFirstTunnelFrameReady(MediaCodec codec) {
+              Log.i(TAG, "onFirstTunnelFrameReady");
```
Done

Jous-Chan commented on starboard/android/apk/app/src/main/java/dev/cobalt/media/MediaCodecBridge.java:
```diff
@@ -531,6 +532,24 @@ public void onFrameRendered(MediaCodec codec, long presentationTimeUs, long nano
+              }
+            }
+          };
+      Log.i(TAG, "setup setOnFirstTunnelFrameReadyListener");
```
Done

Jous-Chan commented on starboard/android/shared/video_decoder.cc:
```diff
@@ -298,7 +299,12 @@ class VideoRenderAlgorithmTunneled : public VideoRenderAlgorithmBase {
   void Render(MediaTimeProvider* media_time_provider,
               std::list<scoped_refptr<VideoFrame>>* frames,
-              VideoRendererSink::DrawFrameCB draw_frame_cb) override {}
+              VideoRendererSink::DrawFrameCB draw_frame_cb) override {
```
Yes, it will be called during tunnel mode. 
If we do not pop those dummy frame (kNeedMoreInput), eos callback will not trigger and ended event will not be fired.

jasonzhangxx replied:
I see, thanks for the clarification.

Jous-Chan commented on starboard/android/shared/video_decoder.cc:
```diff
@@ -871,14 +880,24 @@ void VideoDecoder::WriteInputBuffersInternal(
+      }
+      if (prerolled && !job_token_.is_valid()) {
+        job_token_ =
+            Schedule(std::bind(&VideoDecoder::OnTunnelModePrerollTimeout, this),
```
The original timer is set when input_buffer_written_ == 0. If we enable kForceFlushDecoderDuringReset flag and perform seek, input_buffer_written_ will not be zero. That's why we start another timer here.

Jous-Chan commented on starboard/android/shared/media_decoder.cc:
```diff
@@ -499,13 +502,27 @@ bool MediaDecoder::ProcessOneInputBuffer(
+      if (drm_system_ && input_buffer->drm_info()) {
+        status = media_codec_bridge_->QueueSecureInputBuffer(
+            dequeue_input_result.index, kNoOffset, *input_buffer->drm_info(),
+            pts_us, BUFFER_FLAG_DECODE_ONLY);
```
Pass seek_to_time to MediaCodecBridge.java will add another api in cc/java code.
How about we add another check in queueInputBuffer()/queueSecureInputBuffer() in MediaCodecBridge.java to reset BUFFER_FLAG_DECODE_ONLY? 
Ex:
// BUFFER_FLAG_DECODE_ONLY is added in Android 14
if (Build.VERSION_SDK_INT) < 34) {
    flags &= ~BUFFER_FLAG_DECODE_ONLY;
}

jasonzhangxx replied:
We can still use existing QueueInputBuffer and QueueSecureInputBuffer functions. Comparing to a timestamp of `seek_to_time`, I think a boolean of `is_decode_only` also works and might be better. We then check the boolean in java code and set BUFFER_FLAG_DECODE_ONLY when it's available.

jasonzhangxx commented:
Note there's no new change pushed to this pr, though from the comment it seems that MTK has made changes.

Jous-Chan commented on starboard/android/shared/video_decoder.cc:
```diff
@@ -291,22 +292,32 @@ void StubDrmSessionKeyStatusesChangedFunc(SbDrmSystem drm_system,
+  explicit VideoRenderAlgorithmTunneled(VideoDecoder * video_decoder,
+		                        VideoFrameTracker* frame_tracker)
+      : video_decoder_(video_decoder),
+	frame_tracker_(frame_tracker) {
```
> I see. In that case, with your code, can we remove the code of scheduling OnTunnelModePrerollTimeout in WriteInputBuffers()?

I don't think we can just remove this part. For the test cases Media Playback Quality TotalVideoFrames, it only pushes 25 video frames. So it can not fulfill the prerolled condition. It will trigger OnTunnelModePrerollTimeout with kInitialPrerollTimeout and finish video preroll and pass the test case.

> frame_tracker is passed into VideoRenderAlgorithmTunneled to get the seek time. As the code now calls VideoDecoder::Seek(), we no longer need to pass |frame_tracker| to VideoRenderAlgorithmTunneled, so we can remove frame_tracker_ here.

frame_tracker also needs to report number of dropped frames in GetDroppedFrames(). Do we need to remove frame_tracker here and get number of dropped frames from video_decoder?

> Then call frame_tracker_->Seek() in VideoDecoder::Seek() and use seek_to_time_ in VideoDecoder instead of video_frame_tracker_->seek_to_time().

I can do this and update the commit.

> Also, I think we no longer need VideoFrameTracker::seek_to_time() and that can be removed.

There is a test case for VideoFrameTracker::seek_to_time() in video_frame_tracker_test.cc . Do we need to remove this test case too ?

jasonzhangxx replied:
> I don't think we can just remove this part. For the test cases Media Playback Quality TotalVideoFrames, it only pushes 25 video frames. So it can not fulfill the prerolled condition. It will trigger OnTunnelModePrerollTimeout with kInitialPrerollTimeout and finish video preroll and pass the test case.

> Okay.

> frame_tracker also needs to report number of dropped frames in GetDroppedFrames(). Do we need to remove frame_tracker here and get number of dropped frames from video_decoder?

You're right, I missed that part. Let's still keep `frame_tracker` there.

>  There is a test case for VideoFrameTracker::seek_to_time() in video_frame_tracker_test.cc . Do we need to remove this test case too ?

Yes, let's remove `VideoFrameTracker::seek_to_time()` and its tests.

jasonzhangxx commented on starboard/android/shared/video_decoder.cc:
```diff
@@ -291,22 +292,31 @@ void StubDrmSessionKeyStatusesChangedFunc(SbDrmSystem drm_system,
+
   void Seek(int64_t seek_to_time) override {
-    frame_tracker_->Seek(seek_to_time);
+    video_decoder_->Seek(seek_to_time);
```
We still need to call `frame_tracker_->Seek(seek_to_time);` here to reset `frame_tracker_`.

jasonzhangxx commented:
I'll ask Xiaoming to take a look, too.

chiyuanchan committed:
[android] Add VSP implementation for tunnel mode

Merge PR#5884 for VSP playback (https://github.com/youtube/cobalt/pull/5884)
Use  Android API [AudioTrack.setPlaybackParams] to set playback speed.

Change-Id: Ic2d0f8d605bd44b2ffecb3d30edf056869c072c8
